services/ingestion/subscriber_test.go

```diff
@@ -59,3 +64,412 @@ func Test_Subscribing(t *testing.T) {
+	cdcEvent, evmBlock, blockEvent, err := newBlock(cadenceHeight, hashes)
+	require.NoError(t, err)
+	flowEvent := flow.Event{
+		Type:  string(blockEvent.Etype),
+		Value: cdcEvent,
+	}
+	evmBlockEvents := flow.BlockEvents{
+		BlockID:        flow.Identifier{0x1},
+		Height:         cadenceHeight,
+		BlockTimestamp: time.Now(),
+		Events:         []flow.Event{flowEvent},
+	}
+
+	currentClient.On(
+		"GetEventsForHeightRange",
+		mock.AnythingOfType("context.backgroundCtx"),
+		"A.b6763b4399a888c8.EVM.BlockExecuted",
+		uint64(cadenceHeight),
+		uint64(cadenceHeight),
+	).Return([]flow.BlockEvents{evmBlockEvents}, nil).Once()
+
+	currentClient.On(
+		"GetEventsForHeightRange",
+		mock.AnythingOfType("context.backgroundCtx"),
+		"A.b6763b4399a888c8.EVM.TransactionExecuted",
+		uint64(cadenceHeight),
+		uint64(cadenceHeight),
+	).Return([]flow.BlockEvents{evmTxEvents}, nil).Once()
+
+	currentClient.SubscribeEventsByBlockHeightFunc = func(
+		ctx context.Context,
+		startHeight uint64,
+		filter flow.EventFilter,
+		opts ...access.SubscribeOption,
+	) (<-chan flow.BlockEvents, <-chan error, error) {
+		events := make(chan flow.BlockEvents)
+		errors := make(chan error)
+
+		blockEvents := flow.BlockEvents{
+			BlockID:        flow.Identifier{0x1},
+			Height:         cadenceHeight,
+			BlockTimestamp: time.Now(),
+			Events:         evmTxEvents.Events,
+		}
+
+		// generate single block
+		cdcEvent, _, blockEvent, err := newBlock(cadenceHeight, hashes[:txCount-2])
+		require.NoError(t, err)
+		flowEvent := flow.Event{
+			Type:  string(blockEvent.Etype),
+			Value: cdcEvent,
+		}
+		blockEvents.Events = append(blockEvents.Events, flowEvent)
+
+		go func() {
+			defer close(events)
+
+			for i := startHeight; i <= endHeight; i++ {
+				if i == cadenceHeight {
+					events <- blockEvents
+				} else {
+					events <- flow.BlockEvents{
+						Height: i,
+					}
+				}
+			}
+		}()
+
+		return events, errors, nil
+	}
+
+	client, err := requester.NewCrossSporkClient(
+		currentClient,
+		sporkClients,
+		zerolog.Nop(),
+		flowGo.Previewnet,
+	)
+	require.NoError(t, err)
+
+	subscriber := NewRPCSubscriber(client, 100, flowGo.Previewnet, zerolog.Nop())
+
+	events := subscriber.Subscribe(context.Background(), 1)
+
+	var prevHeight uint64
+
+	for ev := range events {
+		if prevHeight == endHeight {
+			require.ErrorIs(t, ev.Err, errs.ErrDisconnected)
+			break
+		}
+
+		require.NoError(t, ev.Err)
+
+		// this makes sure all the event heights are sequential
+		eventHeight := ev.Events.CadenceHeight()
+		require.Equal(t, prevHeight+1, eventHeight)
+		prevHeight = eventHeight
+
+		if eventHeight == cadenceHeight {
+			assert.Equal(t, evmBlock, ev.Events.Block())
+			for i := 0; i < txCount; i++ {
+				tx := ev.Events.Transactions()[i]
+				assert.Equal(t, hashes[i], tx.Hash())
+			}
+		}
+	}
+
+	// this makes sure we indexed all the events
+	require.Equal(t, uint64(endHeight), prevHeight)
+}
+
+func Test_SubscribingWithRetryOnErrorMultipleBlocks(t *testing.T) {
```

Comment:
can you add above each test a one-liner comment explaining what it tests

Reply:
Updated in https://github.com/onflow/flow-evm-gateway/pull/529/commits/3c9ead3de7e2a226d0244c2860a2aafa8342e7a5